music21/pitch.py

```diff
@@ -4673,62 +4725,66 @@ def updateAccidentalDisplay(
-                    and self.accidental.displayStatus in (False, None)):
-                if self.accidental.name == 'natural':
-                    self.accidental.displayStatus = self._stepInKeySignature(alteredPitches)
+            if (acc is not None
+                    and acc.displayStatus in (False, None)):
+                if acc.name == 'natural':
+                    acc.displayStatus = self._stepInKeySignature(alteredPitches)
                 else:
-                    self.accidental.displayStatus = not self._nameInKeySignature(alteredPitches)
+                    acc.displayStatus = not self._nameInKeySignature(alteredPitches)
 
             # in case display set to True and in alteredPitches, makeFalse
-            elif (self.accidental is not None
-                  and self.accidental.displayStatus is True
+            elif (acc is not None
+                  and acc.displayStatus is True
                   and self._nameInKeySignature(alteredPitches)):
-                self.accidental.displayStatus = False
+                acc.displayStatus = False
 
             # if no accidental or natural but matches step in key sig
             # we need to show or add or an accidental
-            elif ((self.accidental is None or self.accidental.name == 'natural')
+            elif ((acc is None or acc.name == 'natural')
                   and self._stepInKeySignature(alteredPitches)):
-                if self.accidental is None:
-                    self.accidental = Accidental('natural')
-                self.accidental.displayStatus = True
+                none_to_natural()
+                acc.displayStatus = True
             return  # do not search past
 
         # pitches in past... first search if last pitch in measure
         # at this octave contradicts this pitch.  if so then no matter what
         # we need an accidental.
         for i in reversed(range(len(pitchPast))):
+            # check previous in measure.
             thisPPast = pitchPast[i]
             if thisPPast.step == self.step and thisPPast.octave == self.octave:
-                if thisPPast.name != self.name:  # conflicting alters, need accidental and return
-                    if self.accidental is None:
-                        self.accidental = Accidental('natural')
-                    self.accidental.displayStatus = True
+                # conflicting alters, need accidental and return
+                if thisPPast.name != self.name:
+                    none_to_natural()
+                    acc.displayStatus = True
                     return
                 else:  # names are the same, skip this line of questioning
                     break
+
         # nope, no previous pitches in this octave and register, now more complex things...
+        # but first handle 'if-absolutely-necessary' case.
+        if acc and acc.displayType == 'if-absolutely-necessary':
+            acc.displayStatus = False
+            return
```

Comment:
I think this is too early to short-circuit, if `if-absolutely-necessary` also needs to show a necessary natural contradicting the ks. If it's not what it's supposed to do, then maybe reword `only-for-cancelling-in-register`? (not that I love that phrasing). Just that cancelling a ks seems absolutely necessary.

pr:
```python
>>> m = stream.Measure([key.Key('C-'), note.Note('G#'), note.Note('Cn')])
>>> p = stream.Part(m)
>>> m.notes[-1].pitch.accidental.displayType = 'if-absolutely-necessary'
>>> p.show()
```

<img width="392" alt="Screen Shot 2021-10-20 at 10 09 34 PM" src="https://user-images.githubusercontent.com/38668450/138199018-827674d4-97e3-4d72-b60d-2821478d67f6.png">


***
Maybe this means organizing the cases differently so we can short-circuit somewhere.

Reply:
THANKS!